### ISSUE TITLE
euler_from_quaternion returns a triple that does not represent the input rotation at pitch = ±pi/2: 200/200 random cases fail, worst max|R_in - R_back| = 1.986

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#3950]
At `pitch = ±pi/2` (gimbal lock) `euler_from_quaternion` returns a `(roll, pitch, yaw)` triple that **does not represent the input rotation at all** — not merely a different-but-equivalent decomposition. Over 200 random `(roll, yaw)` pairs at `pitch = +pi/2`, reconstructing the rotation from the returned triple fails in **200/200** cases, with a worst-case `max|R_in − R_back| = 1.986` (the entries of a rotation matrix are bounded by 1, so this is a completely different rotation). The control at `|pitch| < pi/4` is `0/200` failures with a worst case of `1.11e-15`.

`euler_from_quaternion` has no gimbal-lock special case (`kornia/geometry/conversions.py:948-960`):

```python
sinr_cosp = 2.0 * (w * x + y * z)
cosr_cosp = 1.0 - 2.0 * (x * x + yy)
roll = sinr_cosp.atan2(cosr_cosp)

sinp = 2.0 * (w * y - z * x)
sinp = sinp.clamp(min=-1.0, max=1.0)
pitch = sinp.asin()

siny_cosp = 2.0 * (w * z + x * y)
cosy_cosp = 1.0 - 2.0 * (yy + z * z)
yaw = siny_cosp.atan2(cosy_cosp)
```

At `pitch = ±pi/2` both `atan2` arguments in each of `roll` and `yaw` cancel to ~0 simultaneously, so `atan2(≈0, ≈0)` returns whatever the floating-point residue dictates — in the case below, `pi/2` for all three components regardless of the input `roll` and `yaw`.

## Repro

```python
import torch
from kornia.geometry.conversions import (
    euler_from_quaternion as e_from_q,
    quaternion_from_euler as q_from_e,
    quaternion_to_rotation_matrix as q2R,
)

def R_of(r, p, y):
    return q2R(torch.stack(q_from_e(*[torch.tensor(a, dtype=torch.float64) for a in (r, p, y)])))

# single case
r0, p0, y0 = 0.1, torch.pi / 2, 0.2
q = q_from_e(*[torch.tensor(a, dtype=torch.float64) for a in (r0, p0, y0)])
back = [float(a) for a in e_from_q(*q)]
print("input  (roll, pitch, yaw) =", [r0, p0, y0])
print("output (roll, pitch, yaw) =", back)
print("max|R_in - R_back|        =", (R_of(r0, p0, y0) - R_of(*back)).abs().max().item())

# 200 random (roll, yaw) at pitch = +pi/2
g = torch.Generator().manual_seed(11)
fails, worst = 0, 0.0
for _ in range(200):
    r = float(torch.rand(1, generator=g) * 2 * torch.pi - torch.pi)
    y = float(torch.rand(1, generator=g) * 2 * torch.pi - torch.pi)
    q = q_from_e(*[torch.tensor(a, dtype=torch.float64) for a in (r, torch.pi / 2, y)])
    b = [float(a) for a in e_from_q(*q)]
    d = (R_of(r, torch.pi / 2, y) - R_of(*b)).abs().max().item()
    worst = max(worst, d)
    fails += d > 1e-6
print(f"\npitch = +pi/2 : {fails}/200 do not reproduce the input rotation, worst max|R_in - R_back| = {worst:.4g}")

# control: |pitch| < pi/4, same generator continued
fails, worst = 0, 0.0
for _ in range(200):
    r = float(torch.rand(1, generator=g) * 2 * torch.pi - torch.pi)
    y = float(torch.rand(1, generator=g) * 2 * torch.pi - torch.pi)
    p = float(torch.rand(1, generator=g) * (torch.pi / 2 - 0.1) - torch.pi / 4)
    q = q_from_e(*[torch.tensor(a, dtype=torch.float64) for a in (r, p, y)])
    b = [float(a) for a in e_from_q(*q)]
    d = (R_of(r, p, y) - R_of(*b)).abs().max().item()
    worst = max(worst, d)
    fails += d > 1e-6
print(f"|pitch| < pi/4: {fails}/200 fail, worst max|R_in - R_back| = {worst:.4g}")
```

The comparison is deliberately made **on the reconstructed rotation matrix**, not on the angle triple: a different triple representing the same rotation would be perfectly acceptable at gimbal lock, and this test would pass. It does not.

## Observed

```
input  (roll, pitch, yaw) = [0.1, 1.5707963267948966, 0.2]
output (roll, pitch, yaw) = [1.5707963267948966, 1.5707963267948966, 1.5707963267948966]
max|R_in - R_back|        = 0.09983341664682817

pitch = +pi/2 : 200/200 do not reproduce the input rotation, worst max|R_in - R_back| = 1.986
|pitch| < pi/4: 0/200 fail, worst max|R_in - R_back| = 1.11e-15
```

## Expected

At gimbal lock the `(roll, yaw)` split is genuinely ambiguous — only the combination `roll ∓ yaw` is determined — so no implementation can recover the original pair. But every implementation *can and should* return **some** triple that reconstructs the same rotation. That is the standard special case:

```python
# |sinp| >= 1  (gimbal lock)
pitch = copysign(pi/2, sinp)
roll  = 0
yaw   = ∓ 2 * atan2(x, w)      # sign chosen by the sign of pitch; picks the determined combination
```

With that, `quaternion_from_euler(*euler_from_quaternion(q))` reproduces `R(q)` at `pitch = ±pi/2` exactly, and the pair remains a mutual inverse over the whole domain instead of only over `|pitch| < pi/2`. Away from the lock the function is already exact (`1.11e-15` in the control above), so the repair is confined to the degenerate branch.

Two pieces of contributing context for whoever picks this up:

- **No existing test would catch it.** `tests/geometry/test_conversions.py::TestEulerFromQuaternion::test_forth_and_back` compares only `.abs()` of each component and carries the comment `# TODO: check hwo to prevent getting inverted angles sometimes`. Nothing in the suite pins the axis assignment, the composition order, or the gimbal-lock behavior. A fix should come with a test that reconstructs the rotation matrix, as the repro above does.
- This is **not** the same defect as `euler_from_quaternion` silently accepting a non-unit
  quaternion (#3953), which has a different cause and a different one-line fix. Every input in the
  repro above is a unit quaternion produced by `quaternion_from_euler`; the failure here is purely
  the missing `|sinp| >= 1` branch.
- The composition this function inverts is `R = Rz(yaw) · Ry(pitch) · Rx(roll)` — extrinsic X→Y→Z about the fixed axes, equivalently intrinsic Z–Y′–X″ — verified against four candidate products (`1.1e-16` for this one vs `0.64` / `0.55` / `0.25` for the alternatives). The docstrings say only "XYZ convention", which does not distinguish these; that documentation gap is being closed in the batch-3 conversions convention-block PR.

## Classification

**Broken output.** The returned triple is not a valid decomposition of the input rotation under any convention — it is a wrong answer, not an alternative one, and it is returned silently. Per the reference-core lifecycle (Part I / Part III §2) that puts it in the class that may land as an ordinary bugfix without waiting for the coordinated Repair & Deprecation window: the only outputs that change are ones that are currently incorrect, so the non-degenerate default path can be preserved byte-identically (#3916-style discipline).

## Environment

kornia 0.9.0rc1 @ `3358ef16`, torch 2.9.1, cpu. Executed from a worktree carrying the batch-3a docs/tests commits on top of `3358ef16`; `euler_from_quaternion` and `quaternion_from_euler` are byte-identical to `origin/main` there (no diff hunk lands in either; function bodies compare AST-identical).

Posted on behalf of @ducha-aiki by Claude (Fable 5).

**Comment by syahra712** [comment]
I'd like to take this one — could you assign me?

I've got a fix ready: at pitch = ±π/2 the shared cos(pitch) factor in the roll and yaw `atan2` arguments collapses to ~0, so each becomes `atan2(0, 0)`. The gimbal-lock branch snaps pitch to ±π/2, pins roll to 0 and folds the single resolved angle into yaw (`yaw = ∓2·atan2(x, w)`), which reconstructs the input rotation to ~1e-16 across both signs. Detection is `cos(pitch) < 8·sqrt(eps)`, so it scales with dtype and catches the float32 case too.

It also flips the pins the test suite already has for this: the strict `xfail` on the round-trip test comes off, the wart becomes a both-signs regression, and the `#3953` wart triple is updated (a scaled quaternion now saturates asin through the new branch — the norm bug itself is untouched). Full `test_conversions.py` passes locally. Happy to open the PR once assigned.

**Comment by ducha-aiki** [comment]
@syahra712 — it is effectively yours already: #3993 is your fix for this issue and it is under review, so no one else should pick this up. What it is waiting on is on the PR side, not here: the changes-requested review from 2026-08-27 has not had a push since, and the branch is now CONFLICTING with `main`. Rebase, address the review, and ping there. Assignee field is unavailable for non-collaborators; this comment is the claim marker.

Posted on behalf of @ducha-aiki by Claude (Fable 5).
